### PR TITLE
CTF Update Again

### PR DESCRIPTION
**Pull request by AyeZeeBB**
CTF Update
Playtest Ready

Location voting has been added.
New location voting UI.
New location voting victory screen.
New voting music and sound FX.
New respawn menu UI.
New class system via playlist vars.
Legend select in the the respawn menu.
Moved team, flag spawn points into location settings.
Moved deathcam pos, ang into location settings.
Fixed alot of bugs.
Cleaned up code.
Overflow Map Added.

### CODE REVIEW

File: vscripts/_remote_functions_mp.gnut [hunk 1]
```diff
@@ -155,10 +155,23 @@ void function RemoteFunctions_Init_Begin()
...
+			Remote_RegisterClientFunction("UpdateMapVotesClient", "int", 0, 255, "int", 0, 255, "int", 0, 255, "int", 0, 255)
+			Remote_RegisterClientFunction("UpdateUIVotingLocationDone", "int", 0, 255)
+			Remote_RegisterClientFunction("UpdateUIVotingLocationTied",  "int", 0, 255, "int", 0, 255)
+			Remote_RegisterClientFunction("ServerCallback_CTF_SetWinnerScreen", "int", 0, 255)
+
 			// CTF: Server to UI
 			Remote_RegisterUIFunction( "OpenCTFRespawnMenu" )
 			Remote_RegisterUIFunction( "CloseCTFRespawnMenu" )
-			
+			Remote_RegisterUIFunction( "OpenCTFVoteMenu" )
```
Comment (IcePixelx) [inline vscripts/_remote_functions_mp.gnut]: Just merge OpenCTFVoteMenu and CloseCTFVoteMenu into 1 Remote call with an integer that then decides behavior via switch case.

This concept can be applied to a fuck ton of function in this gamemode..

File: vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut [hunk 2]
```diff
@@ -120,6 +137,63 @@ bool function ClientCommand_MIL(entity player, array<string> args)
...
+{
+    if (CTF.playersthatvoted.contains(player.GetPlayerName()))
+        return false
+
+    if (!CTF.votingtime)
+        return false
+
+    int mapid = args[0].tointeger()
+
+    if (mapid == 0)
```
Comment (IcePixelx) [inline vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut]: switch case here

Comment (IcePixelx) [reply vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut]: Maybe also check for map name instead?

File: vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut [hunk 3]
```diff
@@ -204,42 +280,50 @@ void function VotingPhase()
...
     }
-    wait CTF_GetVotingTime()
-
-    int choice
 
-    if(CTF.currentmapindex > file.locationSettings.len() - 1)
-        CTF.currentmapindex = 0
-
-    if(CTF.setmap)
+    if (CTF.mapvotes1 > 0 || CTF.mapvotes2 > 0 || CTF.mapvotes3 > 0 || CTF.mapvotes4 > 0)
```
Comment (IcePixelx) [inline vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut]: This def has to go.

File: vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut [hunk 4]
```diff
@@ -329,38 +413,241 @@ void function StartRound()
...
                     }
                     else
                     {
-                        Remote_CallFunction_Replay(player, "ServerCallback_CTF_TeamWon", 99)
+                        //Remote_CallFunction_Replay(player, "ServerCallback_CTF_TeamWon", 99)
                     }
                 }
             }
-            wait 10
+
```
Comment (IcePixelx) [inline vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut:410]: This whole section needs to be re-written cleaner

File: vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut [hunk 5]
```diff
@@ -1308,40 +1653,43 @@ void function _OnPlayerDied(entity victim, entity attacker, var damageInfo)
...
             float damagetaken = DamageInfo_GetDamage( damageInfo )
             Remote_CallFunction_NonReplay(victim, "ServerCallback_CTF_UpdateDamage", 0, damagetaken)
 
-            Remote_CallFunction_NonReplay(victim, "ServerCallback_CTF_OpenCTFRespawnMenu", CTF.bubbleCenter, CTF.IMCPoints, CTF.MILITIAPoints, attacker)
-            
-            float reservedTime = 4// so we dont immediately go to killcam
-            wait reservedTime
+            if (!CTF.votingtime)
+            {
+                Remote_CallFunction_NonReplay(victim, "ServerCallback_CTF_OpenCTFRespawnMenu", CTF.bubbleCenter, CTF.IMCPoints, CTF.MILITIAPoints, attacker)
```
Comment (IcePixelx) [inline vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut]: ServerCallback_CTF_OpenCTFRespawnMenu should also be a ToggleMenu instead.

File: vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut [hunk 6]
```diff
@@ -25,6 +25,16 @@ global function ServerCallback_CTF_RecaptureFlag
 global function ServerCallback_CTF_EndRecaptureFlag
 global function ServerCallback_CTF_ResetFlagIcons
 global function ServerCallback_CTF_SetPointIconHint
+global function ServerCallback_CTF_OpenCTFVoteMenu
```
Comment (IcePixelx) [inline vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut]: Alot of these callbacks can be merged into 1 remote function.

File: vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut [hunk 7]
```diff
@@ -642,14 +660,180 @@ void function waitrespawn(entity player)
...
+    }
+}
+
+void function UpdateUIVotingLocationDone(int mapid)
+{
+    EmitSoundOnEntity( GetLocalClientPlayer(), "UI_PostGame_Level_Up_Pilot" )
+    RunUIScript( "UpdateVotedLocation", file.locationSettings[mapid].name)
+}
+
+void function UpdateUIVotingLocationTied(int mapid, int done)
```
Comment (IcePixelx) [inline vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut]: Switch case here.

File: vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut [hunk 8]
```diff
@@ -642,14 +660,180 @@ void function waitrespawn(entity player)
...
+
+    if (mapid == 254)
+        RunUIScript( "UpdateVotedLocationTied", "")
+    else
+        RunUIScript( "UpdateVotedLocationTied", file.locationSettings[mapid].name)
+}
+
+void function VoteForMap(int mapid)
+{
+    if(hasvoted)
```
Comment (IcePixelx) [inline vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut:1053]: Maybe don't rely on the client to specify if we already voted.

Comment (IcePixelx) [reply vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut:1053]: Handle it server side.

File: vscripts/gamemodes/custom_ctf/sh_gamemode_custom_ctf.nut [hunk 9]
```diff
@@ -349,7 +362,13 @@ vector function GetFlagLocation(LocationSettingsCTF locationSettings, int team)
             if (team == TEAM_MILITIA)
                 spawnorg = <10684, -18468, -3584>
             break
-        
+        case "Overflow":
```
Comment (IcePixelx) [inline vscripts/gamemodes/custom_ctf/sh_gamemode_custom_ctf.nut]: Switch cases for the team.

File: vscripts/ui/CTFMenus/menu_ctf_vote.nut [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+	AddMenuEventHandler( menu, eUIEvent.MENU_NAVIGATE_BACK, OnCTF_NavigateBack )
+
+	AddButtonEventHandler( Hud_GetChild( menu, "MapVote1" ), UIE_CLICK, OnClickMap1 )
+	AddButtonEventHandler( Hud_GetChild( menu, "MapVote2" ), UIE_CLICK, OnClickMap2 )
+	AddButtonEventHandler( Hud_GetChild( menu, "MapVote3" ), UIE_CLICK, OnClickMap3 )
+	AddButtonEventHandler( Hud_GetChild( menu, "MapVote4" ), UIE_CLICK, OnClickMap4 )
+}
+
+void function OnClickMap1( var button )
```
Comment (IcePixelx) [inline vscripts/ui/CTFMenus/menu_ctf_vote.nut]: These can all be merged into 1

**Review by IcePixelx — CHANGES_REQUESTED** [review]
For rexx.

File: vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut (hunk 8 above)
Comment (r-ex) [reply vscripts/gamemodes/custom_ctf/cl_gamemode_custom_ctf.nut:1053]: it is checked server side as well

File: vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut (hunk 2 above)
Comment (r-ex) [reply vscripts/gamemodes/custom_ctf/_gamemode_custom_ctf.nut]: > Maybe also check for map name instead?

no

**Comment by AyeZeeBB** [comment]
ctf is stable and ready togo.

**Pull request merged**